In [1]:
import numpy as np
import pandas as pd

In [3]:
data=pd.read_csv('StudentsPerformance.csv')
data.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group C            some high school  free/reduced   
1  female        group B                 high school  free/reduced   
2  female        group B            some high school  free/reduced   
3  female        group B                some college      standard   
4  female        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none           0             17             10  
1                    none           8             24             23  
2                    none          18             32             28  
3                    none          11             38             32  
4                    none          22             39             33

In [4]:
data.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [50]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# 1. Find out how many males and females participated in the test.

In [10]:
data['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

There are 518 females and 482 males participated in the test.

# 2. What do you think about the student's parental level of education?

In [54]:
data['parental level of education'].mode()

0    some college
dtype: object

Most of the student's parental level of education is 'some college'

# 3. Who scores the most on average for math, reading and writing based on

## ● Gender

In [37]:
male=data['gender']=='male'
male=data.loc[male,:]
female=data['gender']=='female'
female=data.loc[female,:]

In [42]:
if((male['math score'].mean())>(female['math score'].mean())):
    print('Male score the most in average maths score')
elif((male['math score'].mean())<(female['math score'].mean())):
    print('Female score the most in average maths score')

if((male['reading score'].mean())>(female['reading score'].mean())):
    print('Male score the most in average reading score')
elif((male['reading score'].mean())<(female['reading score'].mean())):
    print('Female score the most in average reading score')
    
if((male['writing score'].mean())>(female['writing score'].mean())):
    print('Male score the most in average writing score')
elif((male['writing score'].mean())<(female['writing score'].mean())):
    print('Female score the most in average writing score')

Male score the most in average maths score
Female score the most in average reading score
Female score the most in average writing score


## ● Test preparation course

In [43]:
data['test preparation course'].value_counts()

none         642
completed    358
Name: test preparation course, dtype: int64

In [44]:
none=data['test preparation course']=='none'
none=data.loc[none,:]
completed=data['test preparation course']=='completed'
completed=data.loc[completed,:]

In [45]:
if((none['math score'].mean())>(completed['math score'].mean())):
    print('Those who didnt completed the preparation score the most in average maths score')
elif((none['math score'].mean())<(completed['math score'].mean())):
    print('Those who completed the preparation score the most in average maths score')

if((none['reading score'].mean())>(completed['reading score'].mean())):
    print('Those who didnt completed the preparation score the most in average reading score')
elif((none['reading score'].mean())<(completed['reading score'].mean())):
    print('Those who completed the preparation score the most in average reading score')
    
if((none['writing score'].mean())>(completed['writing score'].mean())):
    print('Those who didnt completed the preparation score the most in average writing score')
elif((none['writing score'].mean())<(completed['writing score'].mean())):
    print('Those who completed the preparation score the most in average writing score')

Those who completed the preparation score the most in average maths score
Those who completed the preparation score the most in average reading score
Those who completed the preparation score the most in average writing score


# 4. What do you think about the scoring variation for math, reading and writing based on


## ● Gender

In [57]:
data.groupby(['gender']).agg(['min','mean','max'])

math score                 reading score                 writing score  \
              min       mean  max           min       mean  max           min   
gender                                                                          
female          0  63.376448  100            17  72.590734  100            10   
male           27  68.821577  100            23  65.545643  100            15   

                        
             mean  max  
gender                  
female  72.467181  100  
male    63.446058  100

● Comparing the range and mean males are more consitant over all the three tests

● Females are scoring more in rading and writing

● Males score more in maths

## ● Test preparation course

In [59]:
data[['test preparation course','gender','math score','reading score','writing score']].groupby(['test preparation course']).agg(['min','mean','max'])

math score                reading score             \
                               min      mean  max           min       mean   
test preparation course                                                      
completed                       23  69.96648  100            37  74.175978   
none                             0  63.78972  100            17  66.417445   

                             writing score                  
                         max           min       mean  max  
test preparation course                                     
completed                100            36  74.684358  100  
none                     100            10  64.457944  100

● Those who completed the test preparation scores more in all the three tests

# 5. The management needs your help to give bonus points to the top 25% of students based on their maths score, so how will you help the management to achieve this.

In [52]:
top_maths=(data.sort_values(by=['math score'],ascending=False)).head(n=250)

In [53]:
top_maths

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
856    male        group E            some high school      standard   
855    male        group E          associate's degree  free/reduced   
854    male        group D            some high school      standard   
853    male        group D          associate's degree  free/reduced   
852    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
856               completed          77             76             77  
855               completed          77             69             68  
854               completed          77             68             69  
853                    none          77             78             73  
852                    none          77             62             64  

[250 rows x 8 columns]

Dataframe 'top_maths' will give the details of top 25% of students based on their maths score